artichoke-backend/src/extn/core/time/subsec.rs

```diff
@@ -0,0 +1,562 @@
+//! Parser for Ruby Time subsecond parameters to help generate `Time`.
+//!
+//! This module implements the logic to parse two optional parameters in the
+//! `Time.at` function call. These parameters (if specified) provide the number
+//! of subsecond parts to add, and a scale of those subsecond parts (millis, micros,
+//! and nanos).
+
+use crate::convert::implicitly_convert_to_int;
+use crate::extn::core::symbol::Symbol;
+use crate::extn::prelude::*;
+
+const NANOS_IN_SECOND: i64 = 1_000_000_000;
+
+const MILLIS_IN_NANO: i64 = 1_000_000;
+const MICROS_IN_NANO: i64 = 1_000;
+const NANOS_IN_NANO: i64 = 1;
+
+#[allow(clippy::cast_precision_loss)]
+const MIN_FLOAT_SECONDS: f64 = i64::MIN as f64;
+#[allow(clippy::cast_precision_loss)]
+const MAX_FLOAT_SECONDS: f64 = i64::MAX as f64;
+const MIN_FLOAT_NANOS: f64 = 0.0;
+#[allow(clippy::cast_precision_loss)]
+const MAX_FLOAT_NANOS: f64 = NANOS_IN_SECOND as f64;
+
+enum SubsecMultiplier {
+    Millis,
+    Micros,
+    Nanos,
+}
+
+impl SubsecMultiplier {
+    #[must_use]
+    const fn as_nanos(&self) -> i64 {
+        match self {
+            Self::Millis => MILLIS_IN_NANO,
+            Self::Micros => MICROS_IN_NANO,
+            Self::Nanos => NANOS_IN_NANO,
+        }
+    }
+}
+
+impl TryConvertMut<Option<Value>, SubsecMultiplier> for Artichoke {
+    type Error = Error;
+
+    fn try_convert_mut(&mut self, subsec_type: Option<Value>) -> Result<SubsecMultiplier, Self::Error> {
+        if let Some(mut subsec_type) = subsec_type {
+            let subsec_symbol = unsafe { Symbol::unbox_from_value(&mut subsec_type, self)? }.bytes(self);
```

Comment:
if we fail to unpack as a symbol, we need to return an `ArgumentError` instead of the `TypeError` this unboxing will return:

```
[3.1.2] > Time.at(12, 15000, "usec")
<internal:timev>:274:in `at': unexpected unit: usec (ArgumentError)
        from (irb):1:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
[3.1.2] > Time.at(12, 15000, Object.new)
<internal:timev>:274:in `at': unexpected unit: #<Object:0x000000010e0e7b80> (ArgumentError)
        from (irb):2:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
```

Reply:
Filed:

- https://github.com/artichoke/artichoke/issues/2095